network/tlskc.go

```diff
@@ -0,0 +1,99 @@
+package network
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"errors"
+	"math/big"
+	"time"
+
+	"net"
+
+	"github.com/dedis/cothority/log"
+)
+
+type TLSCertPEM []byte
+type TLSKeyPEM []byte
+
+// NewTLSCert returns a x509-certificate valid for all CommonNames.
+func NewTLSCert(serial *big.Int, country, org, orgUnit string,
+	validYear int, subjectKeyID []byte,
+	ips []net.IP) *x509.Certificate {
+	return &x509.Certificate{
+		SerialNumber: serial,
+		Subject: pkix.Name{
+			Country:            []string{country},
+			Organization:       []string{org},
+			OrganizationalUnit: []string{orgUnit},
+			CommonName:         "*",
+		},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(validYear, 0, 0),
+		SubjectKeyId:          subjectKeyID,
+		BasicConstraintsValid: true,
+		IsCA:        true,
+		IPAddresses: ips,
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage: x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign |
+			x509.KeyUsageKeyAgreement,
+	}
+}
+
+// NewTLSKC returns a TLSKC-structure with a PEM-encoded certificate and key.
+func NewCertKey(ca *x509.Certificate, keyLen int) (cert TLSCertPEM, key TLSKeyPEM, err error) {
+	if keyLen < 2048 {
+		log.Warn("Small key-length:", keyLen)
+	}
+	priv, _ := rsa.GenerateKey(rand.Reader, keyLen)
+	pub := &priv.PublicKey
+	x509cert, err := x509.CreateCertificate(rand.Reader, ca, ca, pub, priv)
+	if err != nil {
+		return
+	}
+	cert = pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE",
+		Bytes: x509cert})
+	key = pem.EncodeToMemory(&pem.Block{Type: "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(priv)})
+	return
+}
+
+func secureConfig(c *tls.Config) *tls.Config {
+	c.CipherSuites = []uint16{tls.TLS_RSA_WITH_AES_128_CBC_SHA,
+		tls.TLS_RSA_WITH_AES_256_CBC_SHA,
+		tls.TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
+		tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
```

Comment:
Have you tested all supported suites ? I think we should support only the one we generate keys from since we don't have a typical "client -> server" architecture, but I'm newbie in the TLS field. What do you think ?